### PR TITLE
Feature/snap turn smoothing 21270

### PR DESCRIPTION
**Pull request by trentpolack**
Send smoothed orientation from a local client when a snap-turn is done across the network so all remote agents don't see the snap.

# **Testing**
1) Need a persona with an HMD to execute a snap turn. Should still snap.
2) Need another person (or another computer) to witness the snap turn and see a smoothed orientation, instead of a snap.

### CODE REVIEW

**Comment by hifi-gustavo** [comment]
Build console output for commit: d446640bc8803db5d9fd474775ae80ff2ed5ae80 available at: https://gist.github.com/15e3749d50a670f37dafeca280996f9b<br/><br/>Installers available for download at: <br/>Windows: http://cachefly.highfidelity.com/HighFidelity-Beta-PR10336-d446640.exe<br/>OS X: http://cachefly.highfidelity.com/HighFidelity-Beta-PR10336-d446640.dmg

**Comment by hifi-gustavo** [comment]
Build console output for commit: 4d7eeed343a4122d2df885e5fb854eaf8325d164 available at: https://gist.github.com/15e3749d50a670f37dafeca280996f9b<br/><br/>Installers available for download at: <br/>Windows: http://cachefly.highfidelity.com/HighFidelity-Beta-PR10336-4d7eeed.exe<br/>OS X: http://cachefly.highfidelity.com/HighFidelity-Beta-PR10336-4d7eeed.dmg

**Comment by AndrewMeadows** [comment]
QA_FAIL:  I tested this with HMD on windows and observer on linux.  The snap-turn looks ok but the person on HMD is unable to turn their avatar all the way around by looking behind them.  As the HMD is twisted around the avatar's head will move a little bit, then recenter and the avatar will not actually spin around.

**Comment by PhilipRosedale** [comment]
I tested earlier today with the first build, and didn't see a smoothed turn.  I tested by looking at another avatar in an HMD doing snap turns that was floating in the air (so I was not confused by stepping animations).   The other avatar seemed to delay a sec and then turn in a single frame.

**Comment by mittens** [comment]
Yeah, the first build was all mussed up. The build that Andrew tested after we talked was a build where I realized that the orientation wasn't getting sent over the network because, to the local avatar, the rotation hadn't changed, so there was no need to send it. So, if any smoothing is going on now, it forces network transmission.

I'll figure out the other bit with Andrew tomorrow.

File: interface/src/avatar/Avatar.cpp [hunk 1]
```diff
@@ -1397,31 +1361,15 @@ glm::quat Avatar::getUncachedRightPalmRotation() const {
...
-        return;
-    }
-
-    // Whether or not there is an existing smoothing going on, just reset the smoothing timer and set the starting position as the avatar's current position, then smooth to the new position.
-    _smoothOrientationInitial = getOrientation();
-    _smoothOrientationTarget = orientation;
-    _smoothOrientationTimer = 0.0f;
+    // This is the local avatar, no need to handle any position smoothing.
+    AvatarData::setOrientation(orientation);
+    updateAttitude();
```
Comment (AndrewMeadows) [inline interface/src/avatar/Avatar.cpp]: The comment about this being a "local avatar" is no longer true.

File: interface/src/ui/overlays/ModelOverlay.h [hunk 2]
```diff
@@ -39,7 +39,32 @@ class ModelOverlay : public Volume3DOverlay {
...
+
+	// Smoothing data for blending from one position/orientation to another on remote agents.
+	float _smoothPositionTime;
+	float _smoothPositionTimer;
+	float _smoothOrientationTime;
+	float _smoothOrientationTimer;
+	glm::vec3 _smoothPositionInitial;
+	glm::vec3 _smoothPositionTarget;
+	glm::quat _smoothOrientationInitial;
+	glm::quat _smoothOrientationTarget;
```
Comment (AndrewMeadows) [inline interface/src/ui/overlays/ModelOverlay.h]: Why does ModelOverlay require smoothing?

File: libraries/avatars/src/AvatarData.h [hunk 3]
```diff
@@ -781,6 +795,15 @@ public slots:
...
 
+    // Smoothing.
+    const float SMOOTH_TIME_ORIENTATION = 0.15f;
+
+    // Smoothing data for blending from one position/orientation to another on remote agents.
+    float _smoothOrientationTime;
+    float _smoothOrientationTimer;
+    glm::quat _smoothOrientationInitial;
+    glm::quat _smoothOrientationTarget;
+
```
Comment (AndrewMeadows) [inline libraries/avatars/src/AvatarData.h]: No, I don't think we should move the smoothing down into `AvatarData` class.  Instead do the following:

(1) Move smoothing stuff up to `MyAvatar` class.

(2) Modify `AgentData::toByteArray()` where it checks `if (hasAvatarOrientation)` to use `getLocalOrientationForPublicConsumption()` (but come up with a better name) which is declared virtual in `AvatarData` (and implemented to it just call `getLocalOrientation()`) but is overridden in `MyAvatar` to supply the smoothed orientation.

(3) `SpatiallyNestable::_rotationChanged` is protected which means `MyAvatar` class can change it directly, so there is no need for the `SpatiallyNestable::rotationForceChange()` method.  Instead, go ahead and set `_rotationChanged = true` in `MyAvatar` where the smoothed rotation is updated.

In short: track the smoothed rotation in `MyAvatar` and use it (instead of the real rotation) when packing rotation data for an outgoing packet.

File: interface/src/ui/overlays/ModelOverlay.h (hunk 2 above)
Comment (mittens) [inline interface/src/ui/overlays/ModelOverlay.h]: Huh, yeah. Sorry about that. It doesn't, but I mentioned my the HDD my repo was on got corrupted yesterday, so I had to wipe it and I think the clean-up pass got lost in the wipe. Will remedy shortly.

File: interface/src/avatar/Avatar.cpp (hunk 1 above)
Comment (mittens) [inline interface/src/avatar/Avatar.cpp]: That comment is just an artifact of improper cleaning up.

File: libraries/avatars/src/AvatarData.h (hunk 3 above)
Comment (mittens) [inline libraries/avatars/src/AvatarData.h]: Okay, that makes sense. I was just trying to keep stuff related to network transmission in one place, but I can dig this. Should have an updated commit in ~45 minutes.

Thanks much, sir.

**Comment by hifi-gustavo** [comment]
Build console output for commit: 8f93a9a07464500b07675b948f0f1e003a927ab0 available at: https://gist.github.com/15e3749d50a670f37dafeca280996f9b<br/><br/>Installers available for download at: <br/>Windows: http://cachefly.highfidelity.com/HighFidelity-Beta-PR10336-8f93a9a.exe<br/>OS X: http://cachefly.highfidelity.com/HighFidelity-Beta-PR10336-8f93a9a.dmg

**Comment by hifi-gustavo** [comment]
Build console output for commit: 8f34ec4c7ff647a26423cc5c9f05627edbd45571 available at: https://gist.github.com/15e3749d50a670f37dafeca280996f9b<br/><br/>Installers available for download at: <br/>Windows: http://cachefly.highfidelity.com/HighFidelity-Beta-PR10336-8f34ec4.exe<br/>OS X: http://cachefly.highfidelity.com/HighFidelity-Beta-PR10336-8f34ec4.dmg

**Comment by hifi-gustavo** [comment]
Build console output for commit: 6d1fb8a6f82550e601b8b53d55577b0b49ce5433 available at: https://gist.github.com/15e3749d50a670f37dafeca280996f9b<br/><br/>Installers available for download at: <br/>Windows: http://cachefly.highfidelity.com/HighFidelity-Beta-PR10336-6d1fb8a.exe<br/>OS X: http://cachefly.highfidelity.com/HighFidelity-Beta-PR10336-6d1fb8a.dmg

**Comment by mittens** [comment]
Looking into this.

**Comment by hifi-gustavo** [comment]
Build console output for commit: f79e032841bb01e65a1294318ca4a8ce7f24ecc7 available at: https://gist.github.com/15e3749d50a670f37dafeca280996f9b<br/><br/>Installers available for download at: <br/>Windows: http://cachefly.highfidelity.com/HighFidelity-Beta-PR10336-f79e032.exe<br/>OS X: http://cachefly.highfidelity.com/HighFidelity-Beta-PR10336-f79e032.dmg

**Comment by mittens** [comment]
I'm guessing, based on that gist, that this was just a fluke?

**Comment by murillodigital** [comment]
test this please

**Comment by PhilipRosedale** [comment]
@mittens Is this now testable, once we get a build?

**Comment by hifi-gustavo** [comment]
Build console output for commit: f79e032841bb01e65a1294318ca4a8ce7f24ecc7 available at: https://gist.github.com/15e3749d50a670f37dafeca280996f9b<br/><br/>Installers available for download at: <br/>Windows: http://cachefly.highfidelity.com/HighFidelity-Beta-PR10336-f79e032.exe<br/>OS X: http://cachefly.highfidelity.com/HighFidelity-Beta-PR10336-f79e032.dmg

**Comment by mittens** [comment]
@PhilipRosedale Should be. Andrew and I tag-teamed it which means that there were cool guys walking away from explosions. Back to computers.

File: interface/src/avatar/MyAvatar.cpp [hunk 4]
```diff
@@ -264,6 +268,17 @@ QVariant MyAvatar::getOrientationVar() const {
...
+glm::quat MyAvatar::getOrientationOutbound() const {
+    // Allows MyAvatar to send out smoothed data to remote agents if required.
+    if (_smoothOrientationTimer > _smoothOrientationTime) {
+        return (getLocalOrientation());
+    }
+
+    // Smooth the remote avatar movement.
+    float t = _smoothOrientationTimer / _smoothOrientationTime;
+    float l = easeInOutQuad(glm::clamp(t, 0.0f, 1.0f));
+    return (slerp(_smoothOrientationInitial, _smoothOrientationTarget, l));
```
Comment (AndrewMeadows) [inline interface/src/avatar/MyAvatar.cpp]: `l` is a bad variable name, mosly because it can look like a `1` in some fonts.  Use something more descriptive and such as `blend`, `delta`, `coefficient`, etc.

File: interface/src/avatar/MyAvatar.cpp [hunk 5]
```diff
@@ -1840,10 +1862,17 @@ void MyAvatar::updateOrientation(float deltaTime) {
...
     // update body orientation by movement inputs
+	glm::quat initialOrientation = getOrientation();
     setOrientation(getOrientation() * glm::quat(glm::radians(glm::vec3(0.0f, totalBodyYaw, 0.0f))));
 
+	if (snapTurn) {
+		// Whether or not there is an existing smoothing going on, just reset the smoothing timer and set the starting position as the avatar's current position, then smooth to the new position.
+		_smoothOrientationInitial = initialOrientation;
+		_smoothOrientationTarget = getOrientation();
+		_smoothOrientationTimer = 0.0f;
+	}
```
Comment (AndrewMeadows) [inline interface/src/avatar/MyAvatar.cpp]: @mittens, on code inspection it appears there is at least one path by which you can still get a visible "pop" of the avatar on the remote view.  Specifically, you clamp the initial and target orientations here whenever there is a "snap" but... what happens if the avatar is in the middle of smoothing a smooth already?

The smoothing is "fake" right, so `getOrientation()` would always return the real orientation of the avatar.  Your smoothing code will then start at the last target --> there will be a pop to the new start.  Maybe instead you want to use:?
`_smoothOrientationTarget = getOrientationOutbound();`

File: interface/src/avatar/MyAvatar.h [hunk 6]
```diff
@@ -577,6 +579,17 @@ public slots:
...
+    // Basic ease-in-ease-out function for smoothing values.
+    static inline float easeInOutQuad(float lerpValue) {
+        assert(!((lerpValue < 0.0f) || (lerpValue > 1.0f)));
+
+        if (lerpValue < 0.5f) {
+            return (2.0f * lerpValue * lerpValue);
+        }
+
+        return (lerpValue*(4.0f - 2.0f * lerpValue) - 1.0f);
+    }
```
Comment (AndrewMeadows) [inline interface/src/avatar/MyAvatar.h]: Move this out of the class and make it a c-style utility function at the top of _MyAvatar.cpp_.

File: interface/src/ui/overlays/ModelOverlay.cpp [hunk 7]
```diff
@@ -269,13 +274,4 @@ bool ModelOverlay::findRayIntersectionExtraInfo(const glm::vec3& origin, const g
...
-
-void ModelOverlay::locationChanged(bool tellPhysics) {
-    Base3DOverlay::locationChanged(tellPhysics);
-
-    if (_model && _model->isActive()) {
-        _model->setRotation(getRotation());
-        _model->setTranslation(getPosition());
-    }
-}
+}
```
Comment (AndrewMeadows) [inline interface/src/ui/overlays/ModelOverlay.cpp]: Why does `ModelOverlay` need to be changed again?  How is this change related to smoothing the avatar rotation?

Comment (mittens) [inline interface/src/ui/overlays/ModelOverlay.cpp]: It didn't get changed. It was an artifact of a couple of weeks ago when I first talked to you after wondering why "ModelOverlay" wasn't having any effect. This was just removing those artifacts.

(Also, just as a note, unless you know ModelOverlay is in the `src/ui` path, digging through code hierarchies can end up there and you would just assume it's a model representation -- should really be prefixed UiModelOverlay).

File: interface/src/avatar/MyAvatar.cpp (hunk 5 above)
Comment (mittens) [inline interface/src/avatar/MyAvatar.cpp]: Yeah, good plan.

File: interface/src/avatar/MyAvatar.h (hunk 6 above)
Comment (mittens) [inline interface/src/avatar/MyAvatar.h]: I'm not sure I understand the reasoning for this; given that it's an inlined static function that's pretty concise, standard C++ conventions would put it in the header file.

**Comment by hifi-gustavo** [comment]
Build console output for commit: 0a5dfbb9af15bf2373eff2070903826d074bdf0a available at: https://gist.github.com/15e3749d50a670f37dafeca280996f9b<br/><br/>Installers available for download at: <br/>Windows: http://cachefly.highfidelity.com/HighFidelity-Beta-PR10336-0a5dfbb.exe<br/>OS X: http://cachefly.highfidelity.com/HighFidelity-Beta-PR10336-0a5dfbb.dmg

**Comment by hifi-gustavo** [comment]
Build console output for commit: 557a8fffd77615655babf9861b5a5142681c4d85 available at: https://gist.github.com/15e3749d50a670f37dafeca280996f9b<br/><br/>Installers available for download at: <br/>Windows: http://cachefly.highfidelity.com/HighFidelity-Beta-PR10336-557a8ff.exe<br/>OS X: http://cachefly.highfidelity.com/HighFidelity-Beta-PR10336-557a8ff.dmg

**Comment by PhilipRosedale** [comment]
QA Fail:  Testing with two PC's and this build, one standing in front of the other with HMD on, making snap turns.  The turns look identical to those in release, although perhaps delayed?

**Comment by mittens** [comment]
@PhilipRosedale Which build did you try? Andrew and I found something in `557a8ff` that would have caused an issue.

**Comment by hifi-gustavo** [comment]
Build console output for commit: 786838d09237e6a0dcb1e5f97e477e215cb4292b available at: https://gist.github.com/15e3749d50a670f37dafeca280996f9b<br/><br/>Installers available for download at: <br/>Windows: http://cachefly.highfidelity.com/HighFidelity-Beta-PR10336-786838d.exe<br/>OS X: http://cachefly.highfidelity.com/HighFidelity-Beta-PR10336-786838d.dmg

**Comment by Atlante45** [comment]
test this please

**Comment by hifi-gustavo** [comment]
Build console output for commit: 786838d09237e6a0dcb1e5f97e477e215cb4292b available at: https://gist.github.com/82637c84c17adfa173f6e799f9e293b3<br/><br/>Installers available for download at: <br/>Windows: http://cachefly.highfidelity.com/HighFidelity-Beta-PR10336-786838d.exe<br/>OS X: http://cachefly.highfidelity.com/HighFidelity-Beta-PR10336-786838d.dmg

**Comment by mittens** [comment]
@PhilipRosedale @AndrewMeadows Theoretically, this build should have the latest changes that Andrew and I went over.

File: interface/src/avatar/MyAvatar.cpp [hunk 8]
```diff
@@ -1863,13 +1863,13 @@ void MyAvatar::updateOrientation(float deltaTime) {
...
     // update body orientation by movement inputs
-	glm::quat initialOrientation = getOrientation();
+    glm::quat initialOrientation = getOrientationOutbound();
     setOrientation(getOrientation() * glm::quat(glm::radians(glm::vec3(0.0f, totalBodyYaw, 0.0f))));
 
 	if (snapTurn) {
 		// Whether or not there is an existing smoothing going on, just reset the smoothing timer and set the starting position as the avatar's current position, then smooth to the new position.
 		_smoothOrientationInitial = initialOrientation;
-		_smoothOrientationTarget = getOrientation();
+        _smoothOrientationTarget = getOrientation();
```
Comment (AndrewMeadows) [inline interface/src/avatar/MyAvatar.cpp]: Bad indentation here.

Comment (mittens) [inline interface/src/avatar/MyAvatar.cpp]: Sure thing, will remedy. As for the smoothing, I think the current time over which the smooth is done is *way* too short. But I never got a chance to actually see it to know to correct it. The easing is actually just a linear progression in the middle; it's primarily smoothed at the beginning and end of the process.

**Comment by mittens** [comment]
@AndrewMeadows I just fixed what I think was the root of the inconsistencies. Subtle problems are subtle.

**Comment by PhilipRosedale** [comment]
Cool - I don't want to re-test until you guys tell me there is something that is definitely working for you and that I should look at.  It takes time to set it all up on two machines so I can watch the behavior carefully.   Another idea might be to capture a GIF/MP4 or something of what I *should* be seeing.

**Comment by mittens** [comment]
@PhilipRosedale Absolutely; apologies for that. I don't actually have an HMD (nor two computers), so I'm kind of in the blind here once we switched to altering things based on values sent over the network. 

But, yeah, apologies for all the rigmarole that this entails. This task ended up proving to be drastically more non-trivial than it originally seemed.

**Comment by hifi-gustavo** [comment]
Build console output for commit: fc3cce6b738590aae6bd6026fcd6d3bb4c21f648 available at: https://gist.github.com/15e3749d50a670f37dafeca280996f9b<br/><br/>Installers available for download at: <br/>Windows: http://cachefly.highfidelity.com/HighFidelity-Beta-PR10336-fc3cce6.exe<br/>OS X: http://cachefly.highfidelity.com/HighFidelity-Beta-PR10336-fc3cce6.dmg

**Comment by PhilipRosedale** [comment]
Ah - well my gosh that's crazy.  You can't test this at all!   Yikes.

**Comment by mittens** [comment]
@PhilipRosedale Nope, once Andrew and I figured that altering the data transmitted over the network instead of remote agents simulating it themselves, that basically took my ability to test away.

I feel so naked. So cold.

File: interface/src/avatar/MyAvatar.cpp [hunk 9]
```diff
@@ -387,6 +402,16 @@ void MyAvatar::update(float deltaTime) {
...
     _hmdSensorFacingMovingAverage = lerp(_hmdSensorFacingMovingAverage, _hmdSensorFacing, tau);
 
+	if (_smoothOrientationTimer < _smoothOrientationTime) {
+        _rotationChanged = true;
+		_smoothOrientationTimer+= deltaTime;
+
+        if (_smoothOrientationTimer >= _smoothOrientationTime) {
+            // Take it back to astronomical value land.
+            _smoothOrientationTimer = std::numeric_limits<float>::max();
+        }
```
Comment (AndrewMeadows) [inline interface/src/avatar/MyAvatar.cpp]: Remove this check and set-to-max.  It is unnecessary.  It is sufficient when `_smoothOrientationTimer = _smoothOrientationTime + delta`.

Also, get rid of `_smoothOrientationTime` as a data member.  Replace it with `SMOOTH_TIME_ORIENTATION` everywhere.

**Comment by AndrewMeadows** [comment]
QA_FAIL.  I'm still seeing a problem when I test: the turning still appears to "pop" in the remote view, but it is intermittent: sometimes it kinda looks ok but other times it is very glitchy.  I don't see a problem with the code changes here: it should work in theory.  It is as if the update just isn't getting sent, or is not being relayed by the avatar-mixer: I suspect there is some subtle or misunderstood feature/behavior of our code that is preventing this from working, like an overlooked comparison that is bailing on sending updates.

Since mittens lacks an HMD to test I'll probably have to clone his branch and debug it myself.

**Comment by hifi-gustavo** [comment]
Build console output for commit: ca4b5637689ff49912ae2f2de5b8f0303a0e178c available at: https://gist.github.com/15e3749d50a670f37dafeca280996f9b<br/><br/>Installers available for download at: <br/>Windows: http://cachefly.highfidelity.com/HighFidelity-Beta-PR10336-ca4b563.exe<br/>OS X: http://cachefly.highfidelity.com/HighFidelity-Beta-PR10336-ca4b563.dmg

**Comment by hifi-gustavo** [comment]
Build console output for commit: d7ed727faf586024878cb50d47a8373f2277411e available at: https://gist.github.com/15e3749d50a670f37dafeca280996f9b<br/><br/>Installers available for download at: <br/>Windows: http://cachefly.highfidelity.com/HighFidelity-Beta-PR10336-d7ed727.exe<br/>OS X: http://cachefly.highfidelity.com/HighFidelity-Beta-PR10336-d7ed727.dmg

**Comment by hifi-gustavo** [comment]
Build console output for commit: 497bcbf6f2e9d7e647fac3244e1f518b4dd79c41 available at: https://gist.github.com/15e3749d50a670f37dafeca280996f9b<br/><br/>Installers available for download at: <br/>Windows: http://cachefly.highfidelity.com/HighFidelity-Beta-PR10336-497bcbf.exe<br/>OS X: http://cachefly.highfidelity.com/HighFidelity-Beta-PR10336-497bcbf.dmg

**Comment by themelissabrown** [comment]
@AndrewMeadows can this PR be closed?

**Comment by mittens** [comment]
Not yet, I have to reconcile merge conflicts tomorrow morning.

**Comment by themelissabrown** [comment]
@mittens oh, cool. I thought a new PR had been opened for this task. thank you

**Comment by mittens** [comment]
Just merged hifi/master into this PR (lovely timing on the refactor); builds without a problem now. Ensured all changed handed over to Andrew were still present and solid; relying on Andrew's changes for the actual final fix. (CC @themelissabrown).

**Comment by hifi-gustavo** [comment]
Build console output for commit: a3545661dc3c83cc6cd8288cfd1d0ad204010412 available at: https://gist.github.com/15e3749d50a670f37dafeca280996f9b<br/><br/>Installers available for download at: <br/>Windows: http://cachefly.highfidelity.com/HighFidelity-Beta-PR10336-a354566.exe<br/>OS X: http://cachefly.highfidelity.com/HighFidelity-Beta-PR10336-a354566.dmg

**Comment by PhilipRosedale** [comment]
@andrew should I look at this now?

**Comment by AndrewMeadows** [comment]
@PhilipRosedale yes, it is ready for test.  I've reviewed the code and think it is good but will let one other dev review it and add DEV_APPROVED tag since some of the code is stuff I wrote.

**Comment by YuppyPlays** [comment]
Testing 

(I have 2 computers, 1 for Vive HMD and other for Desktop)

Will post results shortly.

**Comment by YuppyPlays** [comment]
Testing complete:

When snap turning:
HMD View: when performing snap it is quick and instantly orientates you to the new snapped position as normal.
Desktop observing: when the HMD user snap turned, it showed a smooth turning of the HMD users model.

**Passed**
